helpers/embeds.js

```diff
@@ -0,0 +1,82 @@
+import { html, render } from 'lit';
+import { findComposedAncestor } from '../helpers/dom.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		await Promise.all(this.#childNodes.map(childNode => childNode.renderNode(options)));
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
```

Comment:
Curious about the nested case with a named `slot`. For example, where would the `slot="header"` be done? I don't think it would be appropriate for a renderer to necessarily assume that the component it is rendering is being placed into a particular slot, since that's a property of the context in which it is being used. If the slot name is a "standard"/"reserved" property in `props`, maybe we could apply it to the element returned here. But would it be appropriate to expect the slot name there?

Reply:
What we talked about doing is encoding that in the placeholder in some way, so the placeholder's data attribute could track whether its children should be slotted or not. So in that sense it becomes the parent component's responsibility to indicate whether its children should be slotted (and what slot name that should have) and providing that information so the framework can slap the slot name on the child element.

Exactly what that looks like is TBD I think. I haven't written any code for it, but because the user profile card uses a slotted button, I was thinking about making a second silly toy renderer for the button and playing around with what that looks like.

Reply:
Sounds like a good thing to try.